### ISSUE TITLE
Feature request: Protection against accidental Cluster deletion

### ISSUE THREAD

**Issue by patrickherrera** [confluentinc/terraform-provider-confluent#79]
Any chance you could add a flag to enable "delete_protection", and as long as that is `true`, the Cluster cannot be deleted.  I feel it is too easy to make a mistake and have Terraform blow away your Cluster - this would at least make it a two-step process and unlikely to happen accidentally.
I'm thinking a similar pattern to that used by AWS RDS Databases (taken from the AWS TF Provider docs):

> `deletion_protection` - (Optional) If the DB instance should have deletion protection enabled. The database can't be deleted when this value is set to true. The default is false.

**Comment by linouk23** [comment]
Thanks for creating an issue @patrickherrera!

That's a great observation! Did you consider using [prevent_destroy](https://www.terraform.io/language/meta-arguments/lifecycle#prevent_destroy) lifecycle attribute?

Let me know if it helps!

**Comment by patrickherrera** [comment]
Thanks, unfortunately that only prevents destructive changes to a resource, which I hope would never apply to a Cluster anyway.  It won't prevent issues where the resource is accidentally removed from the source or some other silly developer error, since the `prevent_destroy` directive will be lost as well and not applied.

**Comment by linouk23** [comment]
> It won't prevent issues where the resource is accidentally removed from the source

Do you mean scenarios where a resource definition is commented out / removed from TF configuration file?

**Comment by patrickherrera** [comment]
Correct.  Or even potentially renamed and Terraform thinks it should destroy/recreate.
It is one of those low-probability/high-impact things that should be relatively easy to implement and save a lot of heartache

**Comment by linouk23** [comment]
@patrickherrera just to follow up, thanks again for the suggestion!

Looks like TF Provider for GCP uses `deletion_protection` that [defaults](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_bigquery_table.go#L931) to `true` for `google_bigquery_table` [resource](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table) too.

I wonder what's your opinion about the default value, make it `true` and have the following UX:

> : On newer versions of the provider, you must explicitly set deletion_protection=false (and run terraform apply to write the field to state) in order to destroy an instance. It is recommended to not set this field (or set it to true) until you're ready to destroy.

or default to `false` as `aws_db_instance` [does](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance).

**Comment by patrickherrera** [comment]
In a way it is a breaking change so I'm not sure if you have a particular policy for that for your module.  I think it is fine to default to `false` so it doesn't surprise existing users, and people can read the changelog and opt-in if they wish

**Comment by linouk23** [comment]
@patrickherrera that makes sense to me, thanks! Yeah I think if we default `deletion_protection` to `false` it won't be a breaking change.

Are there any other existing [resources](https://registry.terraform.io/providers/confluentinc/confluent/latest/docs/guides/sample-project) (except `conluent_kafka_cluster`) where you think adding `deletion_protection` might be useful?

![image](https://user-images.githubusercontent.com/5459870/183323936-0753b711-512b-4ea2-97f7-1bbb5b985447.png)

I'm thinking about `confluent_kafka_topic` and `confluent_connector`.

**Comment by patrickherrera** [comment]
Good question.  If it defaults to `false` then no harm adding it to those extra resources.  Anything that involves potential loss of data makes a lot of sense, so yes, `topics` and `connectors` would be good.

**Comment by linouk23** [comment]
> Or even potentially renamed and Terraform thinks it should destroy/recreate.

btw as a follow up, we investigated the specific use case of a possible recreation (destroy/create) (when updating an attribute that [has got](https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors) `ForceNew: true`) and it seems like `prevents_destroy = true` will help with that. See more details in [this doc](https://docs.google.com/document/d/1VO9v_lglDA2IgUjGWNe1RVrWRopmcpttyMd3OBi7bsE/edit?usp=sharing).

It seems like the use case for using `deletion_protection` attribute is limited to commenting out / deletion of resource definition in TF configuration. Internally we're still figuring out what's the best default value for `deletion_protection` attribute.

**Comment by patrickherrera** [comment]
Renaming the Resource will also cause it to be deleted and recreated, and I have just confirmed that `prevent_destroy` doesn't have an effect here - Terraform happily deleted the old cluster and created a new one with the new name:
```
confluent_kafka_cluster.cluster: Destroying... [id=lkc-XXX]
confluent_kafka_cluster.cluster: Destruction complete after 2s
confluent_kafka_cluster.CC_cluster: Creating...
confluent_kafka_cluster.CC_cluster: Creation complete after 7s [id=lkc-YYY]
```

`prevent_destroy` did stop me running 'terraform destroy' though, and said I had to remove it first.

**Comment by linouk23** [comment]
update: as a quick attempt to mitigate an issue, we added a note about using `prevents_destroy` to our docs:
![image](https://user-images.githubusercontent.com/5459870/185536724-29910556-d6f5-431b-8fa6-ab2700e8bf7c.png)
